### PR TITLE
Centaur BCS

### PR DESCRIPTION
**Pull request by kshakir**

### CODE REVIEW

File: backend/src/main/scala/cromwell/backend/validation/RuntimeAttributesValidation.scala [hunk 1]
```diff
@@ -406,13 +406,13 @@ trait RuntimeAttributesValidation[ValidatedType] {
...
-          womType.coerceRawValue(value).get
-        }
-      }) orElse Option(BadDefaultAttribute(WomString(value.toString)))
+    optionalRuntimeConfig match {
+      case Some(config) if config.hasPath(key) =>
+        val value = config.getValue(key).unwrapped()
+        coercion collectFirst {
+          case womType if womType.coerceRawValue(value).isSuccess => womType.coerceRawValue(value).get
+        } orElse Option(BadDefaultAttribute(WomString(value.toString)))
+      case _ => None
```
Comment (mcovarr) [inline backend/src/main/scala/cromwell/backend/validation/RuntimeAttributesValidation.scala]: the outer `match` has more the air of a `collect`

File: centaur/test_cromwell.sh [hunk 2]
```diff
@@ -24,8 +24,7 @@ Arguments:
     -r    Directory where script is run (defaults to current directory)
     -g    Generate code coverage output for the centaur main classes
     -c    If supplied, the config file to pass to Cromwell
-    -t    If supplied, the file containing a Refresh Token that can be passed into the appropriate options file
-    -s    If supplied, the Google service account json that should be used for various centaur operations
+    -t    If supplied, the timeout for request-plus-response from Centaur to Cromwell
```
Comment (mcovarr) [inline centaur/test_cromwell.sh]: those really weren't being used?

File: docs/backends/BCS.md [hunk 3]
```diff
@@ -200,13 +200,13 @@ It looks like this:
 
 #### autoReleaseJob
 
-Job count may hit the user quota of Alibaba Cloud BatchCompute service if not deleted in time. 
-It's possible to tell backend to delete the related job when a workflow task finishes by set `autoReleaseJob` to 
-`true` like this:
+The Alibaba Cloud BatchCompute service limits the number of simultaneous jobs per user. Jobs created by the backend are
+deleted when the job finishes. However it is possible to tell backend to not delete the job when the call finishes by
```
Comment (mcovarr) [inline docs/backends/BCS.md]: missing a definite article there

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala [hunk 4]
```diff
@@ -122,7 +124,10 @@ case class SqlWorkflowStore(sqlDatabase: WorkflowStoreSqlDatabase) extends Workf
       case (WorkflowStoreState.Running, true) => RestartableRunning.validNel
         // It's in Aborting state and is being restarted
       case (WorkflowStoreState.Aborting, true) => RestartableAborting.validNel
-      case _ => s"Workflow in state $workflowState, and restarted = $restarted cannot be started and should not have been fetched.".invalidNel
+      case _ =>
+        "Workflow %s in state %s and restarted = %b cannot be started and should not have been fetched."
+          .format(workflowStoreEntry.workflowExecutionUuid, workflowState, restarted)
```
Comment (mcovarr) [inline engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala]: why this change?

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileSystemProvider.scala [hunk 5]
```diff
@@ -267,7 +275,11 @@ final case class OssStorageFileSystemProvider(config: OssStorageConfiguration) e
     }
 
     val exists = OssStorageRetry.from(
-      () => ossClient.doesObjectExist(ossPath.bucket, ossPath.key)
+      () => {
+        val request = new GenericRequest(ossPath.bucket, ossPath.key)
+        request.setLogEnabled(false)
```
Comment (mcovarr) [inline filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileSystemProvider.scala]: seems like there should be a thing that takes an ossPath and returns a request with logging disabled

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala (hunk 4 above)
Comment (kshakir) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala]: I keep seeing this in travis. Without the workflow id one has zero clue what workflow has the problem

File: filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileSystemProvider.scala (hunk 5 above)
Comment (kshakir) [reply filesystems/oss/src/main/scala/cromwell/filesystems/oss/nio/OssStorageFileSystemProvider.scala]: Sure. Someday. https://github.com/broadinstitute/cromwell/issues/3526

File: centaur/test_cromwell.sh (hunk 2 above)
Comment (kshakir) [reply centaur/test_cromwell.sh]: They were removed when JES moved its centaur creds from GCS to vault. The docs just weren't updated at the time.

File: centaur/test_cromwell.sh [hunk 6]
```diff
@@ -35,8 +34,9 @@ INITIAL_DIR=$(pwd)
 RUN_DIR=$(pwd)
 TEST_THREAD_COUNT=16
 CENTAUR_SBT_COVERAGE=false
+CROMWELL_TIMEOUT=10s
 
-while getopts ":hb:r:c:p:j:ge:i:" option; do
+while getopts ":hb:r:c:p:j:gt:e:i:" option; do
```
Comment (mcovarr) [inline centaur/test_cromwell.sh]: ah you're right

File: engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala (hunk 4 above)
Comment (mcovarr) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala]: Oh I meant the change from interpolating to formatting, the workflow ID context I get and appreciate. 😄

Comment (kshakir) [reply engine/src/main/scala/cromwell/engine/workflow/workflowstore/SqlWorkflowStore.scala]: Hah! I just find it easier to read lines that don't scroll on infinitely. I still don't think we have an agreed upon exact cutoff yet <sup>*cough* scalafmt someday...</sup>

File: src/bin/travis/resources/bcs_centaur.conf.ctmpl [hunk 7]
```diff
@@ -1,44 +1,85 @@
...
+          # BUT we should also be able to check the hash of the image in OSS using the config supplied credentials
+          # - https://www.alibabacloud.com/help/doc-detail/50452.htm?spm=a3c0i.l31815en.a3.109.50db5139VKk1FK
+          # - https://www.alibabacloud.com/product/oss?spm=a3c0i.7911826.1023975.dproductb1.454c737bJknGYt#resources
+          # Downloading from DockerHub is incredibly slow in Hangzhou. However for portability of WDL, we should use a
+          # new BCS runtime attribute for a `ossDockerRegistry` instead of this format?
+          # docker: "ubuntu/latest oss://broad-test/registry/ubuntu/"
+
+          # If we do allow the above then we should NOT be trying to hash against DockerHub for the BCS backend.
+          # We should test if we can ping the OSS registry using our existing client code.
+          # TODO: Move the above into a ticket. For now, ignore docker.
```
Comment (Horneth) [inline src/bin/travis/resources/bcs_centaur.conf.ctmpl]: I think you did that now ?
